bokehjs/src/coffee/models/plots/plot.coffee

```diff
@@ -88,32 +115,66 @@ class Plot extends LayoutDOM.Model
     if not @height?
       @height = @plot_height
 
+  Object.defineProperty(this.prototype, "plot_canvas", { get: () -> @_plot_canvas })
+
   _doc_attached: () ->
-    @_plot_canvas.attach_document(@document)
+
+    # Add panels for any side renderers
+    # (Needs to be called in _doc_attached, so that panels can attach to the document.)
+    for side in ['above', 'below', 'left', 'right']
+      layout_renderers = @get(side)
+      for r in layout_renderers
+        r.add_panel(side)
+
+    @plot_canvas.attach_document(@document)
 
   add_renderers: (new_renderers...) ->
-    @_plot_canvas.add_renderers(new_renderers...)
+    renderers = @get('renderers')
+    renderers = renderers.concat(new_renderers)
+    @set('renderers', renderers)
 
   add_layout: (renderer, side="center") ->
-    @_plot_canvas.add_layout(renderer, side)
+    @add_renderers(renderer)
+    if side != 'center'
+      if @document?
+        renderer.add_panel(side)
+      @set(side, @get(side).concat([renderer]))
 
   add_glyph: (glyph, source, attrs={}) ->
-    @_plot_canvas.add_glyph(glyph, source, attrs)
+    if not source?
+      source = new ColumnDataSource.Model()
+
+    attrs = _.extend({}, attrs, {data_source: source, glyph: glyph})
+    renderer = new GlyphRenderer.Model(attrs)
+
+    @add_renderers(renderer)
+
+    return renderer
 
   add_tools: (tools...) ->
-    @_plot_canvas.add_tools(tools...)
+    new_tools = for tool in tools
+      if tool.overlay?
+        @add_renderers(tool.overlay)
+
+      if tool.plot?
+        tool
+      else
+        # XXX: this part should be unnecessary, but you can't configure tool.plot
+        # after construting a tool. When this limitation is lifted, remove this code.
+        attrs = _.clone(tool.attributes)
+        attrs.plot = this
+        new tool.constructor(attrs)
 
-  plot_canvas: () ->
-    @_plot_canvas
+    @set(@toolbar.tools, @get("toolbar").tools.concat(new_tools))
```

Comment:
Addint a layer of indirection to access `tools`, breaks semantics of `select()`/`select_one()`. With the existing implementation it's not possible to select tools. Whatever we do, we want not to break symmetry between this and Python APIs.